arch/arm/core/aarch32/mmu/arm_mmu.c

```diff
@@ -0,0 +1,800 @@
+	for (mem_range = 0; mem_range < ARRAY_SIZE(mmu_zephyr_ranges); mem_range++) {
+		
+		pa          = (uint32_t)(mmu_zephyr_ranges[mem_range].start);
+		rem_size    = (uint32_t)(mmu_zephyr_ranges[mem_range].end) - pa;
+		attrs       = mmu_zephyr_ranges[mem_range].attrs;
+		perms_attrs = arm_mmu_convert_attr_flags(attrs);
+
+		if (((uint32_t)&l1_page_table >= pa) && 
+				((uint32_t)&l1_page_table < (pa + rem_size))) {
+			pt_attrs = attrs;
+		}
+
+		while (rem_size > 0) {
+			if (rem_size >= MB(1) && (pa & 0xFFFFF) == 0) {
+				/*
+				 * Remaining area size > 1 MB & matching alignment
+				 * -> map a 1 MB section instead of individual 4 kB
+				 * pages with identical configuration.
+				 */
+				arm_mmu_l1_map_section(pa, pa, perms_attrs);
+				rem_size -= MB(1);
+				pa += MB(1);
+			} else {
+				arm_mmu_l2_map_page(pa, pa, perms_attrs);
+				rem_size -= (rem_size >= KB(4)) ? KB(4) : rem_size;
+				pa += KB(4);
+			}
+		}
+	}
+
+	/* Set up the memory regions defined at the SoC level */
+	for (mem_range = 0; mem_range < mmu_config.num_regions; mem_range++) {
+		pa          = (uint32_t)(mmu_config.mmu_regions[mem_range].base_pa);
+		va          = (uint32_t)(mmu_config.mmu_regions[mem_range].base_va);
+		rem_size    = (uint32_t)(mmu_config.mmu_regions[mem_range].size);
+		attrs       = mmu_config.mmu_regions[mem_range].attrs;
+		perms_attrs = arm_mmu_convert_attr_flags(attrs);
+
+		while (rem_size > 0) {
+			arm_mmu_l2_map_page(va, pa, perms_attrs);
+			rem_size -= (rem_size >= KB(4)) ? KB(4) : rem_size;
+			va += KB(4);
+			pa += KB(4);
+		}
+	}
+
+	/* Clear TTBR1 */
+	__asm__ __volatile__("mcr p15, 0, %0, c2, c0, 1" : : "r"(reg_val));
+
+	/* Write TTBCR: EAE, security not yet relevant, N[2:0] = 0 */
+	__asm__ __volatile__("mcr p15, 0, %0, c2, c0, 2"
+			     : : "r"(reg_val));
+
+	/* Write TTBR0 */
+	reg_val = ((uint32_t)&l1_page_table.entries[0] & ~0x3FFF);
+
+	/*
+	 * Set IRGN, RGN, S in TTBR0 based on the configuration of the
+	 * memory area the actual page tables are located in.
+	 */
+	if (pt_attrs & MATTR_SHARED) {
+		reg_val |= ARM_MMU_TTBR_SHAREABLE_BIT;
+	}
+
+	if (pt_attrs & MATTR_CACHE_OUTER_WB_WA) {
+		reg_val |= (ARM_MMU_TTBR_RGN_OUTER_WB_WA_CACHEABLE <<
+			    ARM_MMU_TTBR_RGN_SHIFT);
+	} else if (pt_attrs & MATTR_CACHE_OUTER_WT_nWA) {
+		reg_val |= (ARM_MMU_TTBR_RGN_OUTER_WT_CACHEABLE <<
+			    ARM_MMU_TTBR_RGN_SHIFT);
+	} else if (pt_attrs & MATTR_CACHE_OUTER_WB_nWA) {
+		reg_val |= (ARM_MMU_TTBR_RGN_OUTER_WB_nWA_CACHEABLE <<
+			    ARM_MMU_TTBR_RGN_SHIFT);
+	}
+
+	if (pt_attrs & MATTR_CACHE_INNER_WB_WA) {
+		reg_val |= ARM_MMU_TTBR_IRGN0_BIT_MP_EXT_ONLY;
+	} else if (pt_attrs & MATTR_CACHE_INNER_WT_nWA) {
+		reg_val |= ARM_MMU_TTBR_IRGN1_BIT_MP_EXT_ONLY;
+	} else if (pt_attrs & MATTR_CACHE_INNER_WB_nWA) {
+		reg_val |= ARM_MMU_TTBR_IRGN0_BIT_MP_EXT_ONLY;
+		reg_val |= ARM_MMU_TTBR_IRGN1_BIT_MP_EXT_ONLY;
+	}
+
+	__asm__ __volatile__("mcr p15, 0, %0, c2, c0, 0" : : "r"(reg_val));
```

Comment:
I think you can use __get_TTBR0() from the cmsis

Reply:
CMSIS integration is my goal, yet, with CMSIS being an external component, I haven't yet figured out where to commit the CPU-specific header core_ca9.h, similar to the already existing core_cr5.h, as CMSIS header inclusion entry point.

Reply:
CMSIS-Core(A) has already been integrated.

See https://github.com/zephyrproject-rtos/cmsis/tree/master/CMSIS/Core_A/Include

Reply:
When looking for the CMSIS, I only looked within zephyrproject-rtos/zephyr and in ARM-software/CMSIS_5, but not zephyrproject-rtos/... D'oh!

Reply:
the MMU code has been modified to use CMSIS whereever supported, the only exceptions being TTBR1 and TTBCR, which aren't covered by cmsis_cp15.h.